### ISSUE TITLE
Update API code to return data about promoted add-ons

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-server#14923]
As part of the Promoted Add-Ons feature, an add-on can now have a promoted add-on category assigned to it, which could be one of a number of values including recommended, verified tier 1, verified tier 2, line, etc. Once these categories are assignable to add-ons, and being used, we will need the API that returns add-on data to be updated to return the promoted category of the add-on. This will effectively replace the `is_recommended` property currently returned. We will need to decide whether to remove that property, and just provide a promoted category property, or whether to return both.

The Search API will also need to be updated to deal with promoted categories, as it also currently just has a `recommended` flag that can be used both as criteria and also as a sort sequence. A separate issue (or issues) will be opened for that once we better understand the requirements.

**Comment by bobsilverberg** [comment]
@eviljeff Have you put any thought into how this might look, now that we have applications as well? I was thinking we'd have to return a promotion category per application. I think there's enough data and (now) existing logic on the server to be able to return an object with one property per application, with either a category or a null. WDYT?

**Comment by eviljeff** [comment]
I guess it'd be: (on the addon response):
```
...
promoted: {
    'firefox': 'recommended',
    'android': 'recommended',
}
...
```
or
```
...
promoted: {
   'recommended': ['firefox', 'android']
}
...
```
The latter is the closest to the back-end model.  Under the current implementation there would only ever be one or zero members of the `promoted` object though (an add-on can only be in one group).

**Comment by bobsilverberg** [comment]
> I guess it'd be: (on the addon response):
> 
> ```
> ...
> promoted: {
>     'firefox': 'recommended',
>     'android': 'recommended',
> }
> ...
> ```
> 
> or
> 
> ```
> ...
> promoted: {
>    'recommended': ['firefox', 'android']
> }
> ...
> ```
> 
> The latter is the closest to the back-end model. Under the current implementation there would only ever be one or zero members of the `promoted` object though (an add-on can only be in one group).

So we cannot currently have?:

> ```
> ...
> promoted: {
>    'recommended': ['firefox'],
>    'line': ['android'],
> }
> ...
> ```
 
I assumed we could.

Also, I feel like the first example, where the `application` is the key, and not the `category`, is better as it's consistent with other parts of the response, such as `current_version.compatibility` and  `categories`, both of which return an object with the `application` as the key.

**Comment by eviljeff** [comment]
The key line I took from the PRD was:

> An add-on can only belong to a single list at any time

so it wouldn't be possible for an add-on to be in both recommended and line lists.

Okay, so an add-on only in the list for desktop would be:
```
...
promoted: {
    'firefox': 'recommended',
}
...
```
and an add-on that isn't promoted for any groups:
```
...
promoted: {}
...
```

**Comment by bobsilverberg** [comment]
> The key line I took from the PRD was:
> 
> > An add-on can only belong to a single list at any time
> 
> so it wouldn't be possible for an add-on to be in both recommended and line lists.

That makes sense, based on that wording, but the fact that we've been asked for a per-app setting makes me suspicious. I think it's worth verifying that assumption. Is the code currently set up to enforce that rule?

> 
> Okay, so an add-on only in the list for desktop would be:
> 
> ```
> ...
> promoted: {
>     'firefox': 'recommended',
> }
> ...
> ```
> 
> and an add-on that isn't promoted for any groups:
> 
> ```
> ...
> promoted: {}
> ...
> ```

I feel like it would be better for it to be:


An add-on only in the list for desktop would be:

```
...
promoted: {
    'android': null,
    'firefox': 'recommended',
}
...
```

and an add-on that isn't promoted for any groups:

```
...
promoted: {
    'android': null,
    'firefox': null,
}
...
```

because that would allow the consumer to assume that there will be a key for any existing application, and wouldn't require special logic to deal with missing keys. Do you know how `current_version.compatibility` and  `categories` work? Do they always provide a key for each application, or do they work similar to what you suggest above?

Although, if your assumption about an add-on only ever being in one list is accurate, then maybe your first structure, which looks like:
```
...
promoted: {
   'recommended': ['firefox', 'android']
}
...
```

might be better, as there would only ever be a need for a single key in `promoted`. But even if that is the case, for now, we probably don't want to assume that that will never change, because things seem to change, over time.

**Comment by eviljeff** [comment]
> I think it's worth verifying that assumption.

You should bring it up in the meeting if you doubt it's current fact-y-ness.

> Is the code currently set up to enforce that rule?

It's enforced at the database level.  It'd need code changes to switch the `OneToOneField` for a `ForeignKey`, and adjust where it's referenced (`addon.promotedaddon` wouldn't be a single `PromotedAddon` instance any longer, it'd be a queryset).  

Signing would become more tricky too - even if Firefox supported an addon being signed under two different promoted groups it wouldn't know which group was supposed to be for which platform.  I don't know if signing it twice as two versions is practical either - we still have support for multiple File per Version in the codebase but it's supposed to be deprecated so some support might have already been removed.

**Comment by eviljeff** [comment]
>  consumer to assume that there will be a key for any existing application, and wouldn't require special logic to deal with missing keys.

That seems like a lot of extra response data when most addons won't be promoted.  Isn't the logic just `addon.promoted[app]` being `undefined`?

**Comment by eviljeff** [comment]
> Do you know how current_version.compatibility and categories work? Do they always provide a key for each application, or do they work similar to what you suggest above?

I think they both only have keys for supported applications, from the code, though I couldn't find an example addon response in the API to confirm (we default to making all the xpis all platform/application)

**Comment by bobsilverberg** [comment]
> > I think it's worth verifying that assumption.
> 
> You should bring it up in the meeting if you doubt it's current fact-y-ness.

I will!

**Comment by bobsilverberg** [comment]
> > consumer to assume that there will be a key for any existing application, and wouldn't require special logic to deal with missing keys.
> 
> That seems like a lot of extra response data when most addons won't be promoted. Isn't the logic just `addon.promoted[app]` being `undefined`?

Fair enough. That makes sense.

**Comment by bobsilverberg** [comment]
@eviljeff One more question: Is everything now in place for this issue to be worked on, or are there outstanding issues that still need to be completed before this is worked on?

**Comment by eviljeff** [comment]
> @eviljeff One more question: Is everything now in place for this issue to be worked on, or are there outstanding issues that still need to be completed before this is worked on?

I hope the model is stable now, though it's possible something will change.  https://github.com/mozilla/addons-server/pull/14988 touches a bunch of stuff though.

**Comment by bobsilverberg** [comment]
> > @eviljeff One more question: Is everything now in place for this issue to be worked on, or are there outstanding issues that still need to be completed before this is worked on?
> 
> I hope the model is stable now, though it's possible something will change. #14988 touches a bunch of stuff I should mention so that could.

Thanks Andrew. I think it makes sense to hold off a bit longer then. There's no huge rush for this.

**Comment by bobsilverberg** [comment]
@eviljeff with https://github.com/mozilla/addons-server/pull/14988 merged, is there anything else we should wait on before starting on this? If it's ready to be worked on, would you like me to try to take it on, or would you like to work on it?

**Comment by eviljeff** [comment]
I think it's ready.  You can take it on if you want.

**Comment by bobsilverberg** [comment]
@eviljeff I started working on this and I am confused. It looks to me like each `addon` can only have one `promotedaddon`, and each `promotedaddon` can only have one `application_id`. If that's the case, how can a single `addon` be promoted for both `firefox` and `android`?

**Comment by eviljeff** [comment]
https://github.com/mozilla/addons-server/blob/3e0b3805f50f2a0ad6eacd641c7d94027f8c1898/src/olympia/promoted/models.py#L14

**Comment by AlexandraMoga** [comment]
I've verified the Addons detail, search and autocomplete APIs on -dev and I was able to see the new `promoted` field being exposed in the Response json object.


Categories exposed by the API for each promoted group:
- for Recommended - `recommended`
- for Strategic - `strategic`
- for Verified - Tier 1 - `sponsored`
- for Verified - Tier 2 - `verified`
- for Spotlight - `spotlight`
- for Line - `line`


Example for a Verified - Tier 1 add-on:

```
"promoted": {
        "apps": [
            "firefox",
            "android"
        ],
        "category": "sponsored"
    }
```

@bobsilverberg I've noticed that the `recommended` status for some older recommended add-ons  is not being reflected in the new `promoted` field for the search API, although the detail API seems to be correctly updated. From what I can tell, the recommended add-ons that were transferred from DiscoveryItems to the new Promoted add-ons model are the ones affected. 

The below requests refer to the same add-on with Id 616933:
detail: https://addons-dev.allizom.org/api/v4/addons/addon/snacork5678 - `recommended` status reflected
search: https://addons-dev.allizom.org/api/v4/addons/search/?q=snacork-kouse-dunkbug - `recommended` status not reflected

Another example for add-on Id 616888:
detail: https://addons-dev.allizom.org/api/v4/addons/addon/auto-tab-discard
search: https://addons-dev.allizom.org/api/v4/addons/search/?q=auto-tab-discard

**Comment by bobsilverberg** [comment]
Thanks @AlexandraMoga. I wonder if we need to run a reindex in order for the search results to be accurate? Are you able to do that? If not I can get to it in a bit.

**Comment by AlexandraMoga** [comment]
I don't think I can run the reindex myself, but, in any case, this shouldn't be too urgent.

**Comment by eviljeff** [comment]
Short of a full reindex, there would need to be a `.save()` called on every existing PromotedAddon to have the data extracted and stored in the index.  (on dev, stage, and prod). Or we run a reindex on dev and stage and have ops run it on prod (once the code is live in those envs).

**Comment by bobsilverberg** [comment]
@AlexandraMoga I have completed a reindex on dev and stage. Can you please check again?

**Comment by AlexandraMoga** [comment]
The search and autocomplete APIs are returning now the `recommended` status in the `promoted` property both on -dev and stage.
